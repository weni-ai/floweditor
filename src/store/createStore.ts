import { applyMiddleware, createStore, Middleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import thunk from 'redux-thunk';

import rootReducer from 'store/rootReducer';
import AppState, { initialState } from 'store/state';

const middlewares: Middleware[] = [thunk];

export default (state: AppState = initialState) => {
  const store = createStore(
    rootReducer,
    state,
    composeWithDevTools(applyMiddleware(...middlewares)),
  );

  return store;
};
