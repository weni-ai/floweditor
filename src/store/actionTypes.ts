import { Type } from 'config/interfaces';
import { FlowDefinition, FlowMetadata } from 'flowTypes';
import Constants from 'store/constants';
import { EditorState } from 'store/editor';
import {
  Asset,
  AssetStore,
  ContactFields,
  RenderNodeMap,
  FlowIssueMap,
  Search
} from 'store/flowContext';
import { NodeEditorSettings } from 'store/nodeEditor';

// Redux action generic
interface DuxAction<T extends Constants, P extends { [key: string]: any } = {}> {
  type: T;
  payload?: P;
}

// Payload types
interface EditorStatePayload {
  editorState: EditorState;
}

interface DefinitionPayload {
  definition: FlowDefinition;
}

interface BaseLanguagePayload {
  baseLanguage: Asset;
}

interface SearchPayload {
  search: Search;
}

interface UpdateMetadataPayload {
  metadata: FlowMetadata;
}

interface LanguagesPayload {
  languages: Asset[];
}

interface UpdateFlowsPayload {
  flows: { uuid: string; name: string }[];
}

interface UpdateDependenciesPayload {
  dependencies: FlowDefinition[];
}

interface UpdateAssetMapPayload {
  assets: AssetStore;
}

interface UpdateNodesPayload {
  nodes: RenderNodeMap;
}

interface UpdateIssuesPayload {
  issues: FlowIssueMap;
}

interface UpdateTypeConfigPayload {
  typeConfig: Type;
}

interface UpdateUserAddingActionPayload {
  userAddingAction: boolean;
}

interface UpdateNodeEditorSettingsPayload {
  settings: NodeEditorSettings;
}

interface UpdateContactFieldsPayload {
  contactFields: ContactFields;
}

// Action types
export type UpdateNodeEditorSettings = DuxAction<
  Constants.UPDATE_NODE_EDITOR_SETTINGS,
  UpdateNodeEditorSettingsPayload
>;

export type UpdateEditorState = DuxAction<Constants.UPDATE_EDITOR_STATE, EditorStatePayload>;

export type UpdateBaseLanguageAction = DuxAction<
  Constants.UPDATE_BASE_LANGUAGE,
  BaseLanguagePayload
>;

export type UpdateSearchAction = DuxAction<Constants.UPDATE_SEARCH, SearchPayload>;

export type UpdateLanguagesAction = DuxAction<Constants.UPDATE_LANGUAGES, LanguagesPayload>;

export type UpdateDefinitionAction = DuxAction<Constants.UPDATE_DEFINITION, DefinitionPayload>;

export type UpdateFlowsAction = DuxAction<Constants.UPDATE_FLOWS, UpdateFlowsPayload>;

export type UpdateMetadataAction = DuxAction<Constants.UPDATE_METADATA, UpdateMetadataPayload>;

export type UpdateAssetsAction = DuxAction<Constants.UPDATE_ASSET_MAP, UpdateAssetMapPayload>;

export type IncrementSuggestedResultNameCountAction = DuxAction<
  Constants.INCREMENT_SUGGESTED_RESULT_NAME_COUNT
>;

export type UpdateNodesAction = DuxAction<Constants.UPDATE_NODES, UpdateNodesPayload>;

export type UpdateIssuesAction = DuxAction<Constants.UPDATE_ISSUES, UpdateIssuesPayload>;

export type UpdateTypeConfigAction = DuxAction<
  Constants.UPDATE_TYPE_CONFIG,
  UpdateTypeConfigPayload
>;

export type UpdateUserAddingActionAction = DuxAction<
  Constants.UPDATE_USER_ADDING_ACTION,
  UpdateUserAddingActionPayload
>;

export type UpdateContactFieldsAction = DuxAction<
  Constants.UPDATE_CONTACT_FIELDS,
  UpdateContactFieldsPayload
>;

export type UpdateTypeConfig = (typeConfig: Type) => UpdateTypeConfigAction;

export type UpdateUserAddingAction = (userAddingAction: boolean) => UpdateUserAddingActionAction;

export type UpdateBaseLanguage = (baseLanguage: Asset) => UpdateBaseLanguageAction;

export type UpdateSearch = (search: Search) => UpdateSearch;

export type IncrementSuggestedResultNameCount = () => IncrementSuggestedResultNameCountAction;

type ActionTypes =
  | UpdateEditorState
  | UpdateNodeEditorSettings
  | UpdateDefinitionAction
  | UpdateFlowsAction
  | UpdateMetadataAction
  | UpdateAssetsAction
  | IncrementSuggestedResultNameCountAction
  | UpdateNodesAction
  | UpdateIssuesAction
  | UpdateTypeConfigAction
  | UpdateUserAddingActionAction
  | UpdateBaseLanguageAction
  | UpdateLanguagesAction
  | UpdateContactFieldsAction
  | UpdateSearchAction;

export default ActionTypes;
