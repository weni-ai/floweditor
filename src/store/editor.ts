import { FlowPosition } from 'flowTypes';
import ActionTypes, { UpdateEditorState } from 'store/actionTypes';
import Constants from 'store/constants';
import { Asset, RenderNode } from 'store/flowContext';

// tslint:disable:no-shadowed-variable
export interface DragSelection {
  startX?: number;
  startY?: number;
  currentX?: number;
  currentY?: number;
}

export interface DebugState {
  showUUIDs: boolean;
}

export interface CanvasPositions {
  [uuid: string]: FlowPosition;
}

export interface Activity {
  nodes: { [uuid: string]: number };
  segments: { [exitToNodeKey: string]: number };
  recentMessages?: { [key: string]: RecentMessage[] };
  is_starting?: boolean;
}

export interface RecentMessage {
  sent: string;
  text: string;
}

export interface Warning {
  name: string;
  type: string;
}

export interface Warnings {
  [uuid: string]: Warning;
}

export interface EditorState {
  currentRevision: number | null;
  simulating: boolean;
  language: Asset | null;
  translating: boolean;
  fetchingFlow: boolean;
  ghostNode: RenderNode | null;
  dragActive: boolean;
  dragStartTime: number;
  dragDownPosition: FlowPosition | null;
  dragNodeUUID: string | null;
  dragGroup: boolean;
  dragSelection: DragSelection | null;
  selectionActive: boolean;
  debug?: DebugState | null;

  // which poptab is popped
  popped: string;

  warnings: Warnings;

  modalMessage?: ModalMessage;
  saving?: boolean;

  // the currently shown activity, can be
  // simulation or live
  activity: Activity;

  // the current live activity
  liveActivity: Activity;

  // interval in millis we should refresh activity
  activityInterval: number;

  // is our page visible or tabbed away
  visible: boolean;

  scrollToNode: string;
  scrollToAction: string;

  currentGuide: string;
  guidingStep: number;

  mouseState: MouseState;
}

export interface ModalMessage {
  title: string;
  body: string;
}

export enum MouseState {
  SELECT = 'select',
  DRAG = 'drag',
  DRAGGING = 'dragging'
}

export const EMPTY_DRAG_STATE: any = {
  dragStartTime: 0,
  dragDownPosition: null,
  dragActive: false,
  dragNodeUUID: null,
  dragGroup: false,
  dragSelection: null
};

// Initial state
export const initialState: EditorState = {
  currentRevision: null,
  simulating: false,
  translating: false,
  language: null,
  fetchingFlow: false,
  dragStartTime: 0,
  dragDownPosition: null,
  dragActive: false,
  dragNodeUUID: null,
  dragGroup: false,
  dragSelection: null,
  ghostNode: null,
  selectionActive: false,
  debug: null,
  warnings: {},

  popped: null,

  activity: { segments: {}, nodes: {} },
  liveActivity: { segments: {}, nodes: {} },
  activityInterval: 5000,
  visible: true,

  scrollToNode: null,
  scrollToAction: null,

  currentGuide: null,
  guidingStep: -1,

  mouseState: MouseState.SELECT
};

// Action Creator
export const updateEditorState = (editorState: EditorState): UpdateEditorState => ({
  type: Constants.UPDATE_EDITOR_STATE,
  payload: {
    editorState
  }
});

/* Reducer */
export const editorState = (editorState: EditorState = initialState, action: ActionTypes) => {
  switch (action.type) {
    case Constants.UPDATE_EDITOR_STATE:
      return action.payload!.editorState;
    default:
      return editorState;
  }
};

// export our reducer
export default editorState;
