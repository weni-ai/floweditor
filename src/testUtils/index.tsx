// TODO: Remove use of Function
// tslint:disable:ban-types
import { ConfigProviderContext, fakePropType } from 'config/ConfigProvider';
import { mount, ReactWrapper, shallow, ShallowWrapper } from 'enzyme';
import { FlowDefinition, FlowEditorConfig } from 'flowTypes';
import mutate, { Query } from 'immutability-helper';
import * as React from 'react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import createStore from 'store/createStore';
import { RenderNodeMap } from 'store/flowContext';
import { getFlowComponents } from 'store/helpers';
import AppState, { initialState } from 'store/state';
import config from 'test/config';
import * as matchers from 'testUtils/matchers';
import { merge, set } from 'utils';

// we need to use require syntax to bust implicit any
const boring: FlowDefinition = require('test/flows/boring.json');

// force our matchers to be read in
const match = matchers;

export interface Resp {
  assets: { [key: string]: any }[];
}

export interface QueryString {
  [key: string]: string;
}

export const contextTypes: { [key: string]: Function } = {
  config: fakePropType,
  store: fakePropType,
  assetService: fakePropType,
};

export const baseState: AppState = mutate(initialState, {
  flowContext: merge({
    definition: require('test/flows/colors.json') as FlowDefinition,
  }),
  editorState: merge({
    language: { id: 'eng', name: 'English' },
  }),
});

const flowEditorConfig: FlowEditorConfig = config;

export const configProviderContext: ConfigProviderContext = {
  config: flowEditorConfig,
};

export const setMock = (
  implementation?: (...args: any[]) => any,
): Query<jest.Mock> => set(jest.fn(implementation));

/**
 * Compose setup method for component tests
 */
export const composeSetup = <P extends {}>(
  Component: React.ComponentClass | React.SFC,
  baseProps: P = {} as any,
  baseDuxState: AppState | Partial<AppState> = baseState,
  baseContext: ConfigProviderContext = configProviderContext,
) => (
  shallowRender: boolean = true,
  propOverrides: Query<P | Partial<P>> = {},
  duxStateOverrides: Query<AppState | Partial<AppState>> = {},
  contextOverrides: Query<
    ConfigProviderContext | Partial<ConfigProviderContext>
  > = {},
  childContextTypeOverrides: { [key: string]: Function } = {},
  children: (JSX.Element | React.ComponentClass | React.SFC)[] = [],
) => {
  const props = mutate(baseProps, propOverrides);
  const store = createStore(mutate(
    baseDuxState,
    duxStateOverrides,
  ) as AppState);
  let context = mutate(baseContext, merge({ store }));

  if (Object.keys(duxStateOverrides).length > 0) {
    context = mutate(context, duxStateOverrides);
  }

  if (Object.keys(contextOverrides).length > 0) {
    context = mutate(context, contextOverrides as ConfigProviderContext);
  }

  const childContextTypes: { [contextProp: string]: Function } = mutate(
    contextTypes,
    childContextTypeOverrides,
  );
  // tslint:disable-next-line:ban-types
  const wrapper = (shallowRender ? (shallow as Function) : (mount as Function))(
    <Component {...props}>{children}</Component>,
    {
      context,
      childContextTypes,
    },
  );

  return {
    wrapper,
    props,
    context,
    instance: wrapper.instance(),
  };
};

export const composeSpy = (obj: Record<string, any> | React.ComponentClass) => (
  instanceMethod: string,
) =>
  jest.spyOn(
    (obj as React.ComponentClass).prototype || obj,
    instanceMethod as any,
  );

/**
 * Wait for promises in queue to resolve
 */
export const flushPromises = () =>
  new Promise((resolve: any) => setImmediate(resolve));

/**
 * Restore spy mocks (distinct from mocks created w/ jest.fn()).
 * Use to declaratively restore multiple spy mocks,
 * otherwise just call .mockRestore() on lone spy.
 */
export const restoreSpies = (...spies: jest.SpyInstance<any>[]) =>
  spies.forEach((spy: any) => spy.mockRestore());

// To-do: type this method's output, can pass it prop generic ingested by getComponentTestUtils
/**
 * NOTE: borrowed from EventBrite: https://github.com/eventbrite/javascript/blob/master/react/testing.md#finding-nodes
 * Finds all instances of components in the rendered `componentWrapper` that are DOM components
 * with the `data-spec` attribute matching `name`.
 * @param {ReactWrapper|ShallowWrapper} componentWrapper - Rendered componentWrapper (result of mount, shallow, or render)
 * @param  {string} snacName - Name of `data-spec` attribute value to find
 * @param {string|Function} (Optional) typeFilter - (Optional) Expected type of the wrappers (defaults to all HTML tags)
 * @returns {ReactWrapper|ReactWrapper[]|ShallowWrapper|ShallowWrapper[]} Matching DOM components
 */
export const getSpecWrapper = (
  componentWrapper: ReactWrapper<{}, {}> | ShallowWrapper<{}, {}>,
  specName: string,
  attributeName: string = 'data-spec',
): any => componentWrapper.find(`[${attributeName}="${specName}"]`);

export const composeDuxState = (
  query: Query<AppState | Partial<AppState>> = {},
  duxState = baseState,
) => mutate(duxState, query);

export const composeComponentTestUtils = <P extends {}>(
  Component: React.ComponentClass | React.SFC | any,
  baseProps: P = {} as any,
  baseDuxState: AppState | Partial<AppState> = baseState,
  baseContext: ConfigProviderContext = configProviderContext,
) => ({
  setup: composeSetup<P>(Component, baseProps, baseDuxState, baseContext),
  spyOn: composeSpy(Component),
});

export const prepMockDuxState = (): {
  testNodes: RenderNodeMap;
  mockDuxState: AppState;
} => {
  const testNodes = getFlowComponents(boring).renderNodeMap;

  return {
    testNodes,
    mockDuxState: {
      ...initialState,
      flowContext: {
        ...initialState.flowContext,
        definition: boring,
        nodes: testNodes,
      },
    },
  };
};

export const createMockStore: Function = configureStore([thunk]);

export const mock = <T extends {}, K extends keyof T>(
  object: T,
  property: K,
  value: T[K],
) => {
  Object.defineProperty(object, property, { get: () => value });
};
