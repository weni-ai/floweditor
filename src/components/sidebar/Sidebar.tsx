import ReactDOM from 'react-dom';
import { fakePropType } from 'config/ConfigProvider';
import * as React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { RenderNode } from 'store/flowContext';
import { createEmptyNode } from 'store/helpers';
import AppState from 'store/state';
import {
  DispatchWithState,
  OnOpenNodeEditor,
  onOpenNodeEditor,
  mergeEditorState,
  MergeEditorState
} from 'store/thunks';

import i18n from 'config/i18n';
import { applyVueInReact } from 'vuereact-combined';
import styles from './Sidebar.module.scss';
// @ts-ignore
import { unnnicModalNext, unnnicToolTip } from '@weni/unnnic-system';
import GuidingSteps from 'components/guidingsteps/GuidingSteps';
import { MouseState } from 'store/editor';

const UnnnicTooltip = applyVueInReact(unnnicToolTip);

const UnnnicModalNext = applyVueInReact(unnnicModalNext, {
  vue: {
    componentWrap: 'div',
    slotWrap: 'div',
    componentWrapAttrs: {
      style: {
        all: '',
        position: 'relative',
        zIndex: 10e2
      }
    }
  }
});

export interface SidebarStoreProps {
  onCopyClick: () => void;
  onCreateNode: (node: RenderNode) => void;
  onMouseStateChange: (mouseState: MouseState) => void;
  selectionActive: boolean;
  nodes: { [uuid: string]: RenderNode };
  onOpenNodeEditor: OnOpenNodeEditor;
  guidingStep: number;
  currentGuide: string;
  mergeEditorState: MergeEditorState;
  mouseState: MouseState;
}

export class Sidebar extends React.PureComponent<SidebarStoreProps, {}> {
  public static contextTypes = {
    config: fakePropType
  };

  private createSendMessageNode(): void {
    if (this.props.guidingStep === 0 && this.props.currentGuide === 'v2') {
      this.props.mergeEditorState({ guidingStep: 1 });
    }

    const emptyNode = createEmptyNode(null, null, 1, this.context.config.flowType);

    this.props.onCreateNode(emptyNode);
  }

  public componentDidMount(): void {
    if (Object.keys(this.props.nodes).length === 0) {
      // this.showGetStartedModal();
    }
  }

  private showGetStartedModal(): void {
    let getStartedModalEl: HTMLDivElement = document.querySelector('#get-started-modal');

    if (!getStartedModalEl) {
      getStartedModalEl = document.createElement('div');
      getStartedModalEl.setAttribute('id', 'get-started-modal');
      document.body.appendChild(getStartedModalEl);
    }

    if (getStartedModalEl.hasChildNodes()) {
      return;
    }

    ReactDOM.render(
      <UnnnicModalNext
        type="alert"
        scheme="feedback-yellow"
        title={i18n.t('empty_flow_message_title')}
        description={i18n.t('empty_flow_message_description')}
        actionPrimaryLabel={i18n.t('buttons.create_message')}
        actionPrimaryButtonType="secondary"
        on={{
          'click-action-primary': () => {
            this.createSendMessageNode();

            ReactDOM.unmountComponentAtNode(getStartedModalEl);
          },
          'click-action-secondary': () => {
            ReactDOM.unmountComponentAtNode(getStartedModalEl);
          }
        }}
        actionSecondaryLabel={i18n.t('buttons.later')}
      />,
      getStartedModalEl
    );
  }

  private getCopyTooltip(): string {
    if (!this.props.selectionActive) {
      return i18n.t('disabled_copy_tooltip');
    }

    return i18n.t('copy_tooltip');
  }

  private getCopyIconScheme(): string {
    if (!this.props.selectionActive) {
      return styles.disabled;
    }

    return styles.enabled;
  }

  private handleCopyClick(): void {
    if (!this.props.selectionActive) {
      return;
    }

    this.props.onCopyClick();
  }

  private toggleMouseState(): void {
    const mouseState =
      this.props.mouseState === MouseState.SELECT ? MouseState.DRAG : MouseState.SELECT;

    this.props.onMouseStateChange(mouseState);
  }

  public render(): JSX.Element {
    return (
      <div className={styles.sidebar}>
        <UnnnicTooltip
          enabled={this.props.guidingStep !== 0}
          side="right"
          text={
            this.props.mouseState === MouseState.SELECT
              ? i18n.t('select', 'Select')
              : i18n.t('drag', 'Drag')
          }
          shortcutText={'V'}
        >
          <div className={styles.option} onClick={() => this.toggleMouseState()}>
            {this.props.mouseState === MouseState.SELECT ? (
              <span className="material-symbols-rounded">near_me</span>
            ) : (
              <span className="material-symbols-rounded">back_hand</span>
            )}
          </div>
        </UnnnicTooltip>

        <GuidingSteps
          guide="v2"
          step={0}
          title={i18n.t('guiding.v2.0.title', 'New block')}
          description={i18n.t(
            'guiding.v2.0.description',
            `Now just click on this button to create a new block.`
          )}
          buttonText={i18n.t('guiding.v2.0.button', 'Got it 1/3')}
        >
          <UnnnicTooltip
            text={i18n.t('create_block')}
            enabled={this.props.guidingStep !== 0}
            side="right"
          >
            <div className={styles.option} onClick={() => this.createSendMessageNode()}>
              <span className="material-symbols-rounded">add_circle</span>
            </div>
          </UnnnicTooltip>
        </GuidingSteps>

        <GuidingSteps
          guide="v2"
          step={1}
          title={i18n.t('guiding.v2.1.title', 'Copy and paste')}
          description={i18n.t(
            'guiding.v2.1.description',
            `Finally, you can copy and paste the blocks wherever you want, including in\nanother flow. Click the button or use the shortcut Ctrl C + Ctrl V.`
          )}
          buttonText={i18n.t('guiding.v2.1.button', 'Got it 2/3')}
        >
          <UnnnicTooltip
            className={styles.left_aligned}
            text={this.getCopyTooltip()}
            enabled={this.props.guidingStep !== 1}
            side="right"
            shortcutText={this.props.selectionActive ? 'Ctrl C' : null}
          >
            <div
              className={`${styles.option} ${!this.props.selectionActive ? styles.disabled : ''}`}
              onClick={() => this.handleCopyClick()}
            >
              <span className={'material-symbols-rounded ' + this.getCopyIconScheme()}>
                content_copy
              </span>
            </div>
          </UnnnicTooltip>
        </GuidingSteps>
      </div>
    );
  }
}

/* istanbul ignore next */
const mapStateToProps = ({
  flowContext: { nodes },
  editorState: { selectionActive, guidingStep, currentGuide }
}: AppState) => {
  return {
    nodes,
    selectionActive,
    guidingStep,
    currentGuide
  };
};

/* istanbul ignore next */
const mapDispatchToProps = (dispatch: DispatchWithState) =>
  bindActionCreators({ onOpenNodeEditor, mergeEditorState }, dispatch);

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Sidebar);
