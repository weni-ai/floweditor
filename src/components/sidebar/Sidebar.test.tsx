import * as React from 'react';
import { render, fireEvent } from 'test/utils';
import { composeComponentTestUtils, setMock } from 'testUtils';
import { Sidebar, SidebarStoreProps } from './Sidebar';
import { MouseState } from '../../store/editor';
import { shallowToJson } from 'enzyme-to-json';

const baseProps: SidebarStoreProps = {
  onCopyClick: vi.fn(),
  onCreateNode: vi.fn(),
  onMouseStateChange: vi.fn(),
  selectionActive: false,
  nodes: {},
  onOpenNodeEditor: vi.fn(),
  guidingStep: 0,
  currentGuide: null,
  mergeEditorState: vi.fn(),
  mouseState: MouseState.SELECT,
  handleSearchChange: vi.fn(),
  search: {
    isSearchOpen: false,
    value: '',
    nodes: [],
    selected: 0,
  },
};

const { setup } = composeComponentTestUtils(Sidebar, baseProps);

describe('Sidebar', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('should render', async () => {
    const { baseElement } = render(<Sidebar {...baseProps} />);

    expect(baseElement).toMatchSnapshot();
  });

  it('should change mouse state on click', async () => {
    const { baseElement, getByTestId } = render(<Sidebar {...baseProps} />);

    fireEvent.click(getByTestId('mouse-state'));
    expect(baseProps.onMouseStateChange).toHaveBeenCalledWith(MouseState.DRAG);

    expect(baseElement).toMatchSnapshot();
  });

  it('should change mouse state to select on click', async () => {
    const newProps = { ...baseProps, mouseState: MouseState.DRAG };
    const { baseElement, getByTestId } = render(<Sidebar {...newProps} />);

    fireEvent.click(getByTestId('mouse-state'));
    expect(baseProps.onMouseStateChange).toHaveBeenCalledWith(
      MouseState.SELECT,
    );

    expect(baseElement).toMatchSnapshot();
  });

  it('should create node on click', async () => {
    const { baseElement, getByTestId } = render(<Sidebar {...baseProps} />);

    fireEvent.click(getByTestId('create-node'));
    expect(baseProps.onCreateNode).toHaveBeenCalledTimes(1);

    expect(baseElement).toMatchSnapshot();
  });

  it('should copy on click', async () => {
    const { baseElement, getByTestId, rerender } = render(
      <Sidebar {...baseProps} />,
    );

    fireEvent.click(getByTestId('copy-button'));
    expect(baseProps.onCopyClick).toHaveBeenCalledTimes(0);

    rerender(<Sidebar {...baseProps} selectionActive={true} />);
    fireEvent.click(getByTestId('copy-button'));
    expect(baseProps.onCopyClick).toHaveBeenCalledTimes(1);

    expect(baseElement).toMatchSnapshot();
  });

  it('should open search on click', () => {
    const component = setup(true, {
      handleSearchChange: setMock(),
    });

    const props: Partial<SidebarStoreProps> = component.props;

    expect(props.handleSearchChange).not.toHaveBeenCalled();
    component.wrapper.find('[data-testid="search-button"]').simulate('click');
    expect(props.handleSearchChange).toHaveBeenCalledTimes(1);
    expect(props.handleSearchChange).toHaveBeenCalledWith({
      isSearchOpen: true,
      value: '',
      nodes: [{}],
      selected: 0,
    });

    expect(shallowToJson(component.wrapper)).toMatchSnapshot();
  });
});
