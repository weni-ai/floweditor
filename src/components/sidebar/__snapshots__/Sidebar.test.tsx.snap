// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Sidebar > should change mouse state on click 1`] = `
<body>
  <div>
    <div
      class="_sidebar_53de86"
    >
      <div
        class="_guiding_wrapper_168ca1"
      >
        <div
          data-testid="veaury-UnnnicTooltip-stub"
        >
          <div
            class="_option_53de86"
            data-testid="mouse-state"
          >
            <div
              data-testid="veaury-Icon-stub"
            />
          </div>
        </div>
      </div>
      <div
        class="_guiding_wrapper_168ca1"
      >
        <div
          data-testid="veaury-UnnnicTooltip-stub"
        >
          <div
            class="_option_53de86"
            data-testid="create-node"
          >
            <div
              data-testid="veaury-Icon-stub"
            />
          </div>
        </div>
      </div>
      <div
        class="_guiding_wrapper_168ca1"
      >
        <div
          data-testid="veaury-UnnnicTooltip-stub"
        >
          <div
            class="_option_53de86 _disabled_53de86"
            data-testid="copy-button"
          >
            <div
              data-testid="veaury-Icon-stub"
            />
          </div>
        </div>
      </div>
      <div
        data-testid="veaury-UnnnicTooltip-stub"
      >
        <div
          class="_option_53de86"
          data-testid="search-button"
        >
          <div
            data-testid="veaury-Icon-stub"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Sidebar > should change mouse state to select on click 1`] = `
<body>
  <div>
    <div
      class="_sidebar_53de86"
    >
      <div
        class="_guiding_wrapper_168ca1"
      >
        <div
          data-testid="veaury-UnnnicTooltip-stub"
        >
          <div
            class="_option_53de86"
            data-testid="mouse-state"
          >
            <div
              data-testid="veaury-Icon-stub"
            />
          </div>
        </div>
      </div>
      <div
        class="_guiding_wrapper_168ca1"
      >
        <div
          data-testid="veaury-UnnnicTooltip-stub"
        >
          <div
            class="_option_53de86"
            data-testid="create-node"
          >
            <div
              data-testid="veaury-Icon-stub"
            />
          </div>
        </div>
      </div>
      <div
        class="_guiding_wrapper_168ca1"
      >
        <div
          data-testid="veaury-UnnnicTooltip-stub"
        >
          <div
            class="_option_53de86 _disabled_53de86"
            data-testid="copy-button"
          >
            <div
              data-testid="veaury-Icon-stub"
            />
          </div>
        </div>
      </div>
      <div
        data-testid="veaury-UnnnicTooltip-stub"
      >
        <div
          class="_option_53de86"
          data-testid="search-button"
        >
          <div
            data-testid="veaury-Icon-stub"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Sidebar > should copy on click 1`] = `
<body>
  <div>
    <div
      class="_sidebar_53de86"
    >
      <div
        class="_guiding_wrapper_168ca1"
      >
        <div
          data-testid="veaury-UnnnicTooltip-stub"
        >
          <div
            class="_option_53de86"
            data-testid="mouse-state"
          >
            <div
              data-testid="veaury-Icon-stub"
            />
          </div>
        </div>
      </div>
      <div
        class="_guiding_wrapper_168ca1"
      >
        <div
          data-testid="veaury-UnnnicTooltip-stub"
        >
          <div
            class="_option_53de86"
            data-testid="create-node"
          >
            <div
              data-testid="veaury-Icon-stub"
            />
          </div>
        </div>
      </div>
      <div
        class="_guiding_wrapper_168ca1"
      >
        <div
          data-testid="veaury-UnnnicTooltip-stub"
        >
          <div
            class="_option_53de86 "
            data-testid="copy-button"
          >
            <div
              data-testid="veaury-Icon-stub"
            />
          </div>
        </div>
      </div>
      <div
        data-testid="veaury-UnnnicTooltip-stub"
      >
        <div
          class="_option_53de86"
          data-testid="search-button"
        >
          <div
            data-testid="veaury-Icon-stub"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Sidebar > should create node on click 1`] = `
<body>
  <div>
    <div
      class="_sidebar_53de86"
    >
      <div
        class="_guiding_wrapper_168ca1"
      >
        <div
          data-testid="veaury-UnnnicTooltip-stub"
        >
          <div
            class="_option_53de86"
            data-testid="mouse-state"
          >
            <div
              data-testid="veaury-Icon-stub"
            />
          </div>
        </div>
      </div>
      <div
        class="_guiding_wrapper_168ca1"
      >
        <div
          data-testid="veaury-UnnnicTooltip-stub"
        >
          <div
            class="_option_53de86"
            data-testid="create-node"
          >
            <div
              data-testid="veaury-Icon-stub"
            />
          </div>
        </div>
      </div>
      <div
        class="_guiding_wrapper_168ca1"
      >
        <div
          data-testid="veaury-UnnnicTooltip-stub"
        >
          <div
            class="_option_53de86 _disabled_53de86"
            data-testid="copy-button"
          >
            <div
              data-testid="veaury-Icon-stub"
            />
          </div>
        </div>
      </div>
      <div
        data-testid="veaury-UnnnicTooltip-stub"
      >
        <div
          class="_option_53de86"
          data-testid="search-button"
        >
          <div
            data-testid="veaury-Icon-stub"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Sidebar > should open search on click 1`] = `
<div
  className="_sidebar_53de86"
>
  <Connect(GuidingSteps)
    buttonText="Got it 1/3"
    description="Click here, use the “V” shortcut or hold down the space bar to use the “little hand” tool, with it you will be able to navigate through your entire flow just by clicking and dragging on the screen"
    guide="control_tools"
    step={0}
    title="Select and drag"
  >
    <Component
      enabled={false}
      shortcutText="V"
      side="right"
      text="Select"
    >
      <div
        className="_option_53de86"
        data-testid="mouse-state"
        onClick={[Function]}
      >
        <Component
          icon="near_me"
          size="md"
        />
      </div>
    </Component>
  </Connect(GuidingSteps)>
  <Connect(GuidingSteps)
    buttonText="Got it 1/3"
    description="Now just click on this button to create a new block."
    guide="v2"
    step={0}
    title="New block"
  >
    <Component
      enabled={false}
      side="right"
      text="Create block"
    >
      <div
        className="_option_53de86"
        data-testid="create-node"
        onClick={[Function]}
      >
        <Component
          icon="add_circle"
          size="md"
        />
      </div>
    </Component>
  </Connect(GuidingSteps)>
  <Connect(GuidingSteps)
    buttonText="Got it 2/3"
    description="Finally, you can copy and paste the blocks wherever you want, including in
another flow. Click the button or use the shortcut Ctrl C + Ctrl V."
    guide="v2"
    step={1}
    title="Copy and paste"
  >
    <Component
      className="_left_aligned_53de86"
      enabled={true}
      shortcutText={null}
      side="right"
      text="Select one or more blocks to copy"
    >
      <div
        className="_option_53de86 _disabled_53de86"
        data-testid="copy-button"
        onClick={[Function]}
      >
        <Component
          className="_disabled_53de86"
          icon="content_copy"
          size="md"
        />
      </div>
    </Component>
  </Connect(GuidingSteps)>
  <Component
    className="_left_aligned_53de86"
    enabled={true}
    shortcutText="Ctrl + F"
    side="right"
    text="Search block"
  >
    <div
      className="_option_53de86"
      data-testid="search-button"
      onClick={[Function]}
    >
      <Component
        icon="search"
        size="md"
      />
    </div>
  </Component>
</div>
`;

exports[`Sidebar > should render 1`] = `
<body>
  <div>
    <div
      class="_sidebar_53de86"
    >
      <div
        class="_guiding_wrapper_168ca1"
      >
        <div
          data-testid="veaury-UnnnicTooltip-stub"
        >
          <div
            class="_option_53de86"
            data-testid="mouse-state"
          >
            <div
              data-testid="veaury-Icon-stub"
            />
          </div>
        </div>
      </div>
      <div
        class="_guiding_wrapper_168ca1"
      >
        <div
          data-testid="veaury-UnnnicTooltip-stub"
        >
          <div
            class="_option_53de86"
            data-testid="create-node"
          >
            <div
              data-testid="veaury-Icon-stub"
            />
          </div>
        </div>
      </div>
      <div
        class="_guiding_wrapper_168ca1"
      >
        <div
          data-testid="veaury-UnnnicTooltip-stub"
        >
          <div
            class="_option_53de86 _disabled_53de86"
            data-testid="copy-button"
          >
            <div
              data-testid="veaury-Icon-stub"
            />
          </div>
        </div>
      </div>
      <div
        data-testid="veaury-UnnnicTooltip-stub"
      >
        <div
          class="_option_53de86"
          data-testid="search-button"
        >
          <div
            data-testid="veaury-Icon-stub"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;
